chrome/content/zotero/xpcom/connector/server_connectorIntegration.js

```diff
@@ -0,0 +1,72 @@
+/*
+	***** BEGIN LICENSE BLOCK *****
+	
+	Copyright © 2017 Center for History and New Media
+					George Mason University, Fairfax, Virginia, USA
+					http://zotero.org
+	
+	This file is part of Zotero.
+	
+	Zotero is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Affero General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+	
+	Zotero is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with Zotero.  If not, see <http://www.gnu.org/licenses/>.
+	
+	***** END LICENSE BLOCK *****
+*/
+
+/**
+ * Adds integration endpoints related to doc integration via HTTP/connector.
+ * 
+ * document/execCommand initiates an integration command and responds with the
+ * next request for the http client (e.g. 'Application.getDocument').
+ * The client should respond to document/respond with the payload and expect
+ * another response with the next request, until it receives 'Document.complete'
+ * at which point the integration transaction is considered complete.
+ */
+Zotero.Server.Endpoints['/connector/document/execCommand'] = function() {};
+Zotero.Server.Endpoints['/connector/document/execCommand'].prototype = {
+	supportedMethods: ["POST"],
+	supportedDataTypes: ["application/json"],
+	permitBookmarklet: true,
+	init: function(data, sendResponse) {
+		if (Zotero.HTTPIntegrationClient.inProgress) {
+			// This will focus the last integration window if present
+			Zotero.Integration.execCommand('http', data.command, data.docId);
+			sendResponse(503, 'text/plain', 'Integration transaction is already in progress')
```

Comment:
Is there a mechanism to avoid needing to restart if the integration client doesn't respond and the session doesn't get closed? I think I ran into that during testing.

Reply:
It was very easy to trigger with the earlier versions and I've made it much more robust since. If any of the integration plugins acts up during the transaction it makes further transactions impossible without a restart. There is no easy way to solve this. Timeouts or multiple sessions are not OK, since the document may be very big and/or take a long time to process.

Reply:
OK, we can see how it goes. If it becomes a problem, it seems like we could maybe do a timeout based on the time of the last field update or something like that.